src/mongo/db/commands/resize_oplog.cpp

```diff
@@ -0,0 +1,111 @@
+
+#include "mongo/platform/basic.h"
+
+#include <string>
+#include <vector>
+
+#include "mongo/db/auth/authorization_manager.h"
+#include "mongo/db/auth/authorization_manager_global.h"
+#include "mongo/db/auth/authorization_session.h"
+#include "mongo/db/catalog/apply_ops.h"
+#include "mongo/db/client.h"
+#include "mongo/db/commands.h"
+#include "mongo/db/commands/dbhash.h"
+#include "mongo/db/concurrency/write_conflict_exception.h"
+#include "mongo/db/db_raii.h"
+#include "mongo/db/dbdirectclient.h"
+#include "mongo/db/jsobj.h"
+#include "mongo/db/matcher/matcher.h"
+#include "mongo/db/operation_context.h"
+#include "mongo/db/repl/oplog.h"
+#include "mongo/db/repl/repl_client_info.h"
+#include "mongo/db/repl/replication_coordinator_global.h"
+#include "mongo/db/catalog/collection.h"
+#include "mongo/db/catalog/collection_catalog_entry.h"
+#include "mongo/db/service_context.h"
+#include "mongo/util/log.h"
+#include "mongo/util/scopeguard.h"
+
+namespace mongo {
+
+using std::string;
+using std::stringstream;
+
+class CmdResizeOplog : public Command {
+public:
+    virtual bool slaveOk() const {
+        return true;
+    }
+    virtual bool supportsWriteConcern(const BSONObj& cmd) const override {
+        return false;
+    }
+    virtual void help(stringstream& help) const {
+        help << "resize oplog size";
+    }
+    CmdResizeOplog() : Command("resizeOplog") {}
+    bool run(OperationContext* opCtx,
+            const string& dbname,
+            BSONObj& jsobj,
+            int,
+            string& errmsg,
+            BSONObjBuilder& result) {
+        StringData dbName("local");
+        AutoGetDb autoDb(opCtx, dbName, MODE_X);
+        Database* const db = autoDb.getDb();
+        Collection* coll = db ? db->getCollection("local.oplog.rs") : nullptr;
+        if (!coll) {
+            return appendCommandStatus(result, Status(ErrorCodes::NamespaceNotFound, "ns does not exist"));
+        }
+        if (!coll->isCapped()) {
+            return appendCommandStatus(result, Status(ErrorCodes::InternalError, "ns does not exist"));
+        }
+        if (!jsobj["size"].isNumber()) {
+            return appendCommandStatus(result, Status(ErrorCodes::InvalidOptions, "invalid size field, size should be a number"));
+        }
+
+        long long size = jsobj["size"].numberLong();
+        if (size < 990LL * 1024 * 1024) {
+            return appendCommandStatus(result, Status(ErrorCodes::InvalidOptions, "oplog size should be 990MB at least"));
+        }
+        WriteUnitOfWork wunit(opCtx);
+        Status status = coll->getRecordStore()->updateCappedSize(opCtx, size);
+        if (!status.isOK()) {
+            return appendCommandStatus(result, status);
+        }
+        CollectionCatalogEntry* entry = coll->getCatalogEntry();
+        entry->updateCappedSize(opCtx, size);
+        wunit.commit();
+        LOG(1) << "resizeOplog success, currentSize:" << size;
```

Comment:
This should probably be at the default log level, it's a rare but significant event